File: wasm_parts/traceql/transpiler/clickhouse_transpiler/aggregator.go
package clickhouse_transpiler

import (
	"strconv"
	"time"
	sql "wasm_parts/sql_select"
	"wasm_parts/traceql/shared"
)

type AggregatorPlanner struct {
	Main       shared.SQLRequestPlanner
	Fn         string
	Attr       string
	CompareFn  string
	CompareVal string

	fCmpVal float64
}

func (a *AggregatorPlanner) Process(ctx *shared.PlannerContext) (sql.ISelect, error) {
	main, err := a.Main.Process(ctx)
	if err != nil {
		return nil, err
	}

	agg, err := a.getAggregator()
	if err != nil {
		return nil, err
	}

	fn, err := getComparisonFn(a.CompareFn)
	if err != nil {
		return nil, err
	}

	err = a.cmpVal()
	if err != nil {
		return nil, err
	}

	return main.AndHaving(fn(agg, sql.NewFloatVal(a.fCmpVal))), nil
}

func (a *AggregatorPlanner) cmpVal() error {
	if a.Attr == "duration" {
		cmpDuration, err := time.ParseDuration(a.CompareVal)
		if err != nil {
			return err
		}
		a.fCmpVal = float64(cmpDuration.Nanoseconds())
		return nil
	}

	var err error
	a.fCmpVal, err = strconv.ParseFloat(a.CompareVal, 64)
	return err
}

func (a *AggregatorPlanner) getAggregator() (sql.SQLObject, error) {
	switch a.Fn {
	case "count":
		return sql.NewRawObject("toFloat64(count(distinct index_search.span_id))"), nil
	case "avg":
		return sql.NewRawObject("avgIf(agg_val, isNotNull(agg_val))"), nil
	case "max":
		return sql.NewRawObject("maxIf(agg_val, isNotNull(agg_val))"), nil
	case "min":
		return sql.NewRawObject("minIf(agg_val, isNotNull(agg_val))"), nil
	case "sum":
		return sql.NewRawObject("sumIf(agg_val, isNotNull(agg_val))"), nil
	}
	return nil, &shared.NotSupportedError{"aggregator not supported: " + a.Fn}
}

File: traceql/clickhouse_transpiler/aggregator.js
const Sql = require('@cloki/clickhouse-sql')
const { getCompareFn, durationToNs } = require('./shared')

module.exports = class Builder {
  constructor () {
    this.main = null
    this.fn = ''
    this.attr = ''
    this.compareFn = ''
    this.compareVal = ''
  }

  /**
   *
   * @param main {BuiltProcessFn}
   * @returns {Builder}
   */
  withMain (main) {
    this.main = main
    return this
  }

  /**
   *
   * @param fn {string}
   * @returns {Builder}
   */
  withFn (fn) {
    this.fn = fn
    return this
  }

  /**
   *
   * @param attr {string}
   * @returns {Builder}
   */
  withAttr (attr) {
    this.attr = attr
    return this
  }

  /**
   *
   * @param fn {string}
   * @returns {Builder}
   */
  withCompareFn (fn) {
    this.compareFn = fn
    return this
  }

  /**
   *
   * @param val {string}
   * @returns {Builder}
   */
  withCompareVal (val) {
    this.compareVal = val
    return this
  }

  /**
   * @returns {ProcessFn}
   */
  build () {
    const self = this
    /** @type {BuiltProcessFn} */
    const res = (ctx) => {
      const sel = this.main(ctx)
      const fCmpVal = self.cmpVal()
      const agg = self.aggregator()
      const compareFn = getCompareFn(self.compareFn)
      const comparreExp = compareFn(agg, Sql.val(fCmpVal))
      // .having is broken
      sel.having_conditions = Sql.And([...sel.having_conditions.args, comparreExp])
      return sel
    }
    return res
  }

  cmpVal () {
    if (this.attr === 'duration') {
      return durationToNs(this.compareVal)
    }
    return parseFloat(this.compareVal)
  }

  aggregator () {
    switch (this.fn) {
      case 'count':
        return new Sql.Raw('toFloat64(count(distinct index_search.span_id))')
      case 'avg':
        return new Sql.Raw('avgIf(agg_val, isNotNull(agg_val))')
      case 'max':
        return new Sql.Raw('maxIf(agg_val, isNotNull(agg_val))')
      case 'min':
        return new Sql.Raw('minIf(agg_val, isNotNull(agg_val))')
      case 'sum':
        return new Sql.Raw('sumIf(agg_val, isNotNull(agg_val))')
    }
  }
}
